### PR TITLE
Small first iteration for profiler: use nonblocking queue

### PR DESCRIPTION
**Pull request by cjolivier01**
## Description ##

Use nonblocking queue for profiler
Next up: buffer server so allocation not done in-line

### Essentials ###
- [ ] Passed code style checking (`make lint`)
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage
- [ ] For user-facing API changes, API doc string has been updated. For new C++ functions in header files, their functionalities and arguments are well-documented. 
- [ ] To my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [ ] Feature1, tests, (and when applicable, API doc)
- [ ] Feature2, tests, (and when applicable, API doc)

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Interesting edge cases to note here

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -38,7 +38,7 @@ mxnet_option(USE_MKL_EXPERIMENTAL "Use experimental MKL (if MKL enabled and foun
 mxnet_option(USE_OPERATOR_TUNING  "Enable auto-tuning of operators" ON AND NOT MSVC)
 mxnet_option(USE_GPERFTOOLS       "Build with GPerfTools support (if found)" ON)
 mxnet_option(USE_JEMALLOC         "Build with Jemalloc support"   ON)
-mxnet_option(USE_PROFILER         "Build with Profiler support"   OFF)
+mxnet_option(USE_PROFILER         "Build with Profiler support"   ON)
```
Comment (cjolivier01) [inline CMakeLists.txt]: No overhead when not turned on at runtime

File: src/engine/profiler.h [hunk 2]
```diff
@@ -65,11 +66,24 @@ struct OprExecStat {
  */
 struct DevStat {
   /*! \brief device name */
-  std::string dev_name;
+  std::string dev_name_;
   /*! \brief operation execution statistics on this device */
-  std::vector<OprExecStat*> opr_exec_stats;
-  /*! \brief internal mutex of the execution state */
-  std::mutex m_;
```
Comment (cjolivier01) [inline src/engine/profiler.h]: This mutex is no longer required (removed)

**Comment by cjolivier01** [comment]
This build has a dependency on: https://github.com/dmlc/dmlc-core/pull/334

File: src/engine/profiler.cc [hunk 3]
```diff
@@ -56,14 +55,13 @@ Profiler::Profiler()
 #endif
 
   this->profile_stat = new DevStat[cpu_num_ + gpu_num_ + 1];
```
Comment (rahul003) [inline src/engine/profiler.cc]: Should this be deleted in the destructor?

File: src/engine/profiler.cc [hunk 4]
```diff
@@ -113,15 +111,12 @@ OprExecStat *Profiler::AddOprStat(int dev_type, uint32_t dev_id) {
       idx = cpu_num_ + gpu_num_;
       break;
     default:
-      LOG(FATAL) << "Unkown dev_type";
+      LOG(FATAL) << "Unknown dev_type: " << dev_type;
       return NULL;
```
Comment (rahul003) [inline src/engine/profiler.cc]: I know there's a log(fatal) here but Coverity had pointed out that memory is not freed here. Should we delete opr_stat in the default case?

**Comment by cjolivier01** [comment]
Ok

File: src/engine/profiler.cc (hunk 4 above)
Comment (cjolivier01) [reply src/engine/profiler.cc]: Also added deleting in DumpProfile()

File: src/engine/profiler.cc (hunk 3 above)
Comment (cjolivier01) [reply src/engine/profiler.cc]: That's for a different PR.  I actually tried to once, but Profiler is currently a static object right now and things get ugly when it shuts down and tries to delete this.  I will probably be refactoring some of this eventually.

**Comment by cjolivier01** [comment]
Build dependency: https://github.com/dmlc/dmlc-core/pull/341

**Pull request merged**